modules/opengl/src/image/imagecompositor.cpp

```diff
@@ -30,88 +30,51 @@
 #include <modules/opengl/image/imagecompositor.h>
 
 #include <modules/opengl/texture/textureunit.h>
-#include <modules/opengl/texture/texture2d.h>
 #include <modules/opengl/texture/textureutils.h>
 #include <modules/opengl/shader/shaderutils.h>
 #include <modules/opengl/image/imagegl.h>
 #include <modules/opengl/openglutils.h>
-#include <inviwo/core/util/assertion.h>
 
 namespace inviwo {
 
-ImageCompositor::ImageCompositor(std::string programFileName)
-    : shader_(programFileName)
-    , colorTex_({1u, 1u}, GLFormat(), GL_LINEAR)
-    , depthTex_({1u, 1u}, GLFormat(), GL_LINEAR)
-    , pickingTex_({1u, 1u}, GLFormat(), GL_LINEAR) {}
+ImageCompositor::ImageCompositor(std::string programFileName) : shader(programFileName) {}
 
-void ImageCompositor::composite(const Image& source, Image& target, ImageType type) {
-    copyTextures(target);
-
-    utilgl::activateTarget(target, type);
-    shader_.activate();
-    utilgl::GlBoolState depthTest(GL_DEPTH_TEST, false);
+void ImageCompositor::composite(const Image& source0, const Image& source1, Image& destination,
+                                ImageType type) {
+    utilgl::GlBoolState depthTest(GL_DEPTH_TEST, true);
+    utilgl::activateTarget(destination, type);
+    shader.activate();
 
     TextureUnitContainer units;
-    utilgl::bindAndSetUniforms(shader_, units, source, "tex0", ImageType::ColorDepthPicking);
-    bindTextures(units, "tex1");
-    utilgl::setShaderUniforms(shader_, target, "outportParameters");
+    utilgl::bindAndSetUniforms(shader, units, source0, "tex0", ImageType::ColorDepthPicking);
+    utilgl::bindAndSetUniforms(shader, units, source1, "tex1", ImageType::ColorDepthPicking);
+
+    utilgl::setShaderUniforms(shader, destination, "outportParameters");
     utilgl::singleDrawImagePlaneRect();
 
-    shader_.deactivate();
+    shader.deactivate();
     utilgl::deactivateCurrentTarget();
 }
 
-void ImageCompositor::composite(ImageInport& source, ImageOutport& target, ImageType type) {
-    if (source.isReady() && target.hasData()) {
-        copyTextures(*target.getData());
-
-        utilgl::activateTarget(target, type);
-        shader_.activate();
-        utilgl::GlBoolState depthTest(GL_DEPTH_TEST, false);
+void ImageCompositor::composite(const ImageInport& source0, const ImageInport& source1,
+                                ImageOutport& destination, ImageType type) {
+    if (source0.isReady() && source1.isReady()) {
```

Comment:
check whether ports are different (write to same texture as reading is undefined behavior)